### ISSUE TITLE
Have KeyBase publish itself as a public KeyServer

### ISSUE THREAD

**Issue by EliW** [keybase/keybase-issues#327]
It would be an important step in KeyBase adoption IMO, for it to publish itself as a valid keyserver using the standard protocols for such.

As it stands, since keybase doesn't, it means that all the tooling (such as EnigMail for Thunderbird) that has existed for a decade+ ... is unusable against keybase published keys.

**Comment by segfault** [comment]
:+1:

**Comment by sekjun9878** [comment]
+1

**Comment by victorhaggqvist** [comment]
+1

**Comment by wsargent** [comment]
+1

**Comment by jlb1234** [comment]
I agree with @timbray and @ramsey that PGP key management via keyservers and keysigning/web-of-trust has failed, more or less, entirely miserably. I don't even need to get into why.

That doesn't mean that the public key system is broken. Just the way in which they are disseminated and trusted. That's where Keybase can step in. Enigmail and GPGTools can also be built to hook into Keybase instead of standard keyservers.

The matter at hand is, should Keybase be built to interact with this system or should it be more or less separate? I would say that a break with the past could be provide a lot of opportunities for innovation, while interactivity/backwards compatibility is probably not worthwhile except for the small numbers of people who have been using encrypted mail until now - as opposed to the much higher number of people who have never even tried to use it.

**Comment by EliW** [comment]
Except jlb1234 -- By not having interactivity/backwards compatibility ... it means that all the staunch advocates of PKE in the past, who have continued to use it and preach it's wonders in the past ... are excluded from Keybase.  Even if the backwards compat wouldn't be used but by a 'small portion'** of users.  Those users are the ones who have shown consistently that they care and are vocal.  Those are absolutely the users that you want on your side.

** Given that ~30 people above by my counts here have given this a +1, I doubt that the portion is 'all that small' in the end.

**Comment by gimsieke** [comment]
+1

**Comment by wsargent** [comment]
> Given that ~30 people above by my counts here have given this a +1, I doubt that the portion is 'all that small' in the end.

As one of the people who voted +1, I'd like to say I want an end to the cruft, and I want fresh new implementations of key servers that aren't bogged down with 20 years of history.

**Comment by malgorithms** [comment]
Hi all - just so you know, we care about this. It's just unclear how to proceed, and we'll wait until Keybase is further along to address this.

There are 2 separate issues:

1. does keybase accept keys from other servers?
2. does it push to them?

for point 1, it's not simple. A key according to keybase requires a provable connection to an identity somewhere. We're not an email->PGP key service, like the others. At least not until there's a convenient way for someone to prove they own an email address, publicly. (This is discussed elsewhere in this repo.)

Plus, unlike the traditional key servers, we're striving for a canonical ordering of key announcements. Your keybase signature chain is a series of announcements: adding a proof, removing a proof, adding another proof, then another, adding a bitcoin address, things like that. Each of these links in the chain points at the previous. The goals of this system are important: (1) preventing the server from lying by omission, and (2) making sure the server gives the same answer to everyone in the world.

Contrast this with a traditional key server, which could withhold whatever it wants.

I'm not saying this totally prevents us from gossiping keys around, but it does show that keys we bring in need to be treated in a way that has nothing to do with keybase.

As for point 2, pushing keys to other servers, this interests us and should be simpler. Eventually.

Also, one last thing: it requires very few API calls to keybase to mirror our entire database and prove to an outsider that you've mirrored it accurately. This means that anyone could implement a layer that pulls keys from Keybase and pushes them to other servers or just acts as a lookup mirror. Just saying :-)

This latter point can be useful for people who worry about metadata leakage. If you don't want a powerful adversary to know you need foobar's public key, it might be nice to maintain/use a server that you trust that mirrors keybase and checks all the proofs for you. 

The TL;DR here is we don't have anything planned in the short term. Pushing is possible in the medium term.

Also, I have found myself paraphrasing this a bunch:

> As much as I like this idea, as someone who has already implemented a key server and tried to host sks, onak and at least one more - don't underestimate this, running a keyserver is even more horrible than using gpgme or whatever bindings you use. Don't underestimate the effort. ;)

**Comment by maxtaco** [comment]
(And BTW, there's a discussion on #1266 on how to go about mirroring).

**Comment by ramsey** [comment]
I thought this was an interesting read about the problems with PGP/GPG, and I felt it was relevant to this conversation: http://www.thoughtcrime.org/blog/gpg-and-me/

**Comment by malgorithms** [comment]
I recommend this:

http://blog.cryptographyengineering.com/2014/08/whats-matter-with-pgp.html

**Comment by iamale** [comment]
@ramsey
The problems Moxie describes in his post are **exactly** the problems Keybase is trying to solve, don't you think?

**Comment by ramsey** [comment]
@iamale, I didn't say they were different. I was just sharing, since I thought it was relevant.

**Comment by TehShrike** [comment]
I have friends using Enigmail, I would really like to be able to give them a key server they could use to download my public key.

**Comment by iamale** [comment]
I think we could look from the other point of view and implement Keybase support in the Eigmail itself, don't you think?
https://github.com/iamale/enigmail-keybase

**Comment by iamale** [comment]
(Another workaround could be to make a special keyserver that auto-imports keys of people you're tracking on Keybase, thus you can hook it up to any PGP-enabled mail client)

**Comment by DrAzraelTod** [comment]
@iamale 
>  implement Keybase support in the Enigmail itself

yes.. and then every other keyserver-client out there... great idea

**Comment by iamale** [comment]
@DrAzraelTod Well, there shouldn't be _too_ many _popular_ clients out there. Also,

> Another workaround could be to make a special keyserver that auto-imports keys of people you're tracking on Keybase

**Comment by shinji257** [comment]
I was looking at a way to interface with the api and pull public keys automatically.

Using a json parser you can use API call fetch and feed it pgp key ids.  Checking against at least Mailvelope I was able to get my own key successfully.  I used an app called jq to check and see if it was something that was viable.

If a key was found it will give a status result of 0 and a name of OK (I think you can check for either?
At that point the public key will be part of the keys array called bundle.

For jq I used the following filters
.name.status
.keys[].bundle

At this point it is just a matter of stripping the leading/trailing quotes and converting the \n's to actual line breaks.

The response appears to be signed but that's undocumented.  I'm assuming it is with keybase's own key and it is to verify the response.

If the key doesn't exist it throws a code 901 and name KEY_NOT_FOUND.

**Comment by towynlin** [comment]
:+1:

**Comment by terokorp** [comment]
+1

**Comment by troyengel** [comment]
I was just invited to Keybase and this was the very first thing I went looking for, a way to use 'gpg --recv-key' just like everything else (EnigMail, etc.). I want to like Keybase, but guys in order to succeed you have to at least have *basic* interop with how clients verify signatures from new (unknown) email contacts.

+1

**Comment by wsargent** [comment]
Any chance of using opmsg instead? https://github.com/stealth/opmsg

**Comment by iamale** [comment]
@wsargent I doubt. Keybase itself relies heavily on PGP.

**Comment by AymericLabs** [comment]
+1
Been using gpg for a long while. ability to receive public keys with a traditional method is a must for me. 
I may be in the 0.0000001 % some are talking about, and understand it is a complex issue. But E won't promote keybase if it refuses basic compatibility with what I am familiar with, unless there are good reasons to do so. Without it, it is just PKI infrastructure out there.. Your interfaces  are very advanced, big plus for that. I like the canonical approach, and the "make it simple for everyone" approach. I just don't get why not implementing basic standards helps in this objective.

**Comment by Carlgo11** [comment]
+1

**Comment by schock** [comment]
I'm setting up keybase right now, and as soon as I had it set up, this was the first thing I tried to do (open Thunderbird and try to get keybase to play with enigmail). So, +1 both to keybase hosting a keyserver, and to integrating w/enigmail and any other top, most popular, mail encryption tools.

**Comment by Chankster** [comment]
+1

**Comment by zQueal** [comment]
> Keybase itself relies heavily on PGP.

This is a bit misleading, IMO. I've never really seen Keybase identify as anything other than an interface for PGP.

**Comment by vext01** [comment]
+1

**Comment by mabe-at** [comment]
+1

**Comment by dhuseby** [comment]
+1 If keybase really wants to make a better workflow for PKI, then interop is a must.  It seems like the obvious first set of keybase users are people *already* using PGP.  But those people need to have a migration path or they won't make the jump.  Providing a standard sks keyserver interface to keybase.io seems like an obvious requirement to me.

**Comment by dhuseby** [comment]
I just told a friend to forget about keybase.io, you can't use them with any of the existing encrypted email tools.

**Comment by profan** [comment]
+1

**Comment by dtiersch** [comment]
> you can't use them with any of the existing encrypted email tools

is plainly wrong. Counterexample: [OpenKeychain](https://play.google.com/store/apps/details?id=org.sufficientlysecure.keychain&hl=de)

**Comment by shinji257** [comment]
Realistically if someone wanted to an individual could write a shim program that would emulate a standard gpg key server and redirect requests through the keybase api.  Earlier I already showed that it is possible to search the database via the api using information typically used when searching standard pgp servers.

**Comment by iamale** [comment]
> write a shim program that would emulate a standard gpg key server

Yeah, I wanted to at first. But well, it's a bit _harder_ than you might think

**Comment by ariscop** [comment]
i don't believe there's any way to publish your email address on keybase, otherwise it would be easy, at least if you limit it to people you're tracking

**Comment by ariscop** [comment]
See  #490

**Comment by iamale** [comment]
```py
import requests

class KeyserverException(Exception): pass

def add_key(username):
    r = requests.get("https://keybase.io/{username}/key.asc".format(
          username=username))
    r.raise_for_status()
    keytext = r.text

    r = requests.post("https://pgp.mit.edu/pks/add", data=dict(
          keytext=keytext))
    r.raise_for_status()
    if "added to key server database" not in r.text:
        raise KeyserverException(r.text)
```

Then just walk some tree (the Merkle tree? just the “tracking” graph?) and add _all the keys_ to your keyserver of choice. Not the best solution, but seems OK to me for now.

**Comment by ariscop** [comment]
@iamale or import into your local gpg ksystore, but that's not very useful without email addresses

**Comment by iamale** [comment]
@ariscop There are emails in some of the keys (i. e. in mine), while in others the `<user>@keybase.io` address is used. We could modify the identity part with the email taken from https://keybase.io/_/api/1.0/user/lookup.json?username=ale&fields=emails (if any) before importing.

**Comment by eighthave** [comment]
It seems the simple detail that is missing from this conversation is that keybase should implement the standard keyserver API for fetching keys, not anything else.  Then all of the existing clients (which have orders of magnitude more users than the PGP clients that support keybase) will be able to fetch public keys from keybase.  I see no need for keybase to implement the whole keyserver syncing stuff, for example, or the publishing/uploading API.

**Comment by vonpupp** [comment]
:+1: for sure

**Comment by brainwane** [comment]
The [Enigmail bug "Support key lookup via keybase.io"](http://sourceforge.net/p/enigmail/bugs/410/) is marked as fixed as of 24 September 2015 -- I thus imagine that keybase.io lookup will be in the next Enigmail release, although I don't know when that will be, and I'm not brave enough to run git master. (https://github.com/iamale/enigmail-keybase hasn't been committed to since 2012, as far as I can tell; I think the relevant changes are ["added support for keybase protocol (using keybase API) for searching and downloading keys"](http://sourceforge.net/p/enigmail/source/ci/fa046b7d68e039bcc2643f54830bdb9f84f551e3/) and ["added support for importing keys from URL"](http://sourceforge.net/p/enigmail/source/ci/729bca6159ab63493b9532dd3ab2d0b7c165841b/).)

I use Enigmail and Thunderbird, and I'm looking forward using its UI to fetch keys from keybase.io. Maybe other popular clients can follow Enigmail's approach to start, until and unless keybase.io implements a backward-compatible keyserver? (Many users would probably be fine with searching by fingerprint and doing without username or email search, at least to start.)

**Comment by iamale** [comment]
@brainwane

> https://github.com/iamale/enigmail-keybase hasn't been committed to since 2012, as far as I can tell;

Yeah, I've been working on it silently, but there wasn't any success. I've deleted the repo now to reduce confusion. Sorry for that!

**Comment by mitar** [comment]
+1

**Comment by dashohoxha** [comment]
> Without being able to "find" keys on Keybase.io using tools such as EnigMail,
> it is really useless at this point.
> However, this doesn't mean that Keybase needs to become a keyserver, 
> if they push keys to already existing keyservers.

This says it all. It's enough to **push the public keys to the existing keyservers**. And this should not be very difficult. 

In my opinion, it had to be one of the first features of keybase. Not implementing this cannot be justified; it has only one explanation: hating existing keyservers and wanting to undermine them.

Even if the existing public key infrastructure is so bad, you don't through the old shoes without buying yet the new ones. However I don't think that it is that bad. I see this infrastructure as serving two functions: the first is as a key directory (you search by name or email address and you get the key, or vice-versa), and the second one is to (indirectly) support the web-of-trust.

What has failed in so many years is that we don't have yet a world-wide, high-quality, trustable web-of-trust, that can be used to facilitate the everyday communications. But this is not a problem of GPG and public-key servers, it is mostly a social problem. If there is a software failure, it is the GPG clients that have failed (to be easy to understand and use), not the keyservers. So, there is no reason to hate the keyservers and to undermine them.

**Comment by mitar** [comment]
> It's enough to push the public keys to the existing keyservers. And this should not be very difficult. 

This could probably be done even as a community project? It should not be too hard to monitor the stream of all Keybase keys and push them automatically to existing keyservers? Anyone wanna take to implement it? We could probably find a place to host it once is done.

**Comment by brainwane** [comment]
I know that some people prefer not to have their keys on keyservers; is there any conceivable reason a person would be ok with having their key on keybase.io but not be okay with having that key syndicated to keyservers?

Also, if we copy keys to a keyserver, ought we follow CRUD principles and also ensure there's a way to update and delete keys from those keyservers as well if the owner updates or deletes their keys from keybase.io?

**Comment by mitar** [comment]
> is there any conceivable reason a person would be ok with having their key on keybase.io but not be okay with having that key syndicated to keyservers?

Hm, they are in the public blockchain already. :-) Not sure if keyservers are really much worse then. It is just a different infrastructure, but public as well. But good question.

> Also, if we copy keys to a keyserver, ought we follow CRUD principles and also ensure there's a way to update and delete keys from those keyservers as well if the owner updates or deletes their keys from keybase.io?

I do not think you can really delete the key from keyservers. You can revoke them though. But that will have to be first done at Keybase.

**Comment by chigh** [comment]
I don't have mine published on a key server and don't want them there.

**Comment by mitar** [comment]
> I don't have mine published on a key server and don't want them there.

But you do have them on Keybase? Can you elaborate why do you feel like that? It could help provide some insight.

**Comment by jlb1234** [comment]
> But you do have them on Keybase? Can you elaborate why do you feel like that? It could help provide some insight.

My sense is that the problem with Keyservers are twofold: (1) they are easily crawled by bad/malicious actors, and (2) it's almost impossible to remove a key from the Keyservers network due to their decentralized network.

Personally I believe it's important for Keybase to integrate -- at least to some extent -- with the existing public key infrastructure. With that being said I do understand Keybase's founders' (very public and forward) issues with the entire Keyserver infrastructure, including the web of trust, which to an extent KB is created with the express purpose of replacing. 

The fundamental problem of Keyservers - which KB still has not resolved -- is that it's not possible to truly prove that a key is connected to someone's email address. That's why there are 1000s of 'keys' posted for obama@whitehouse.gov. Public key encryption is mathematically sound but you can't really be sure that you are using the right key which opens you up to MITM attacks.

**Comment by zQueal** [comment]
> (2) it's almost impossible to remove a key from the Keyservers network due to their decentralized network.

This is entirely by design -- so to call it a problem is pretty, well, problematic. If a key is revoked a record needs to be maintained outside of "hey, this key is revoked."

> Personally I believe it's important for Keybase to integrate -- at least to some extent -- with the existing public key infrastructure.

This, again, is by design. Keybase seeks to make working with PGP easier and to centralize assertions of identity. It doesn't pretend to be a keyserver. If you're that focused on pushing your public key to keyservers then simply use GPG;

    gpg --keyserver certserver.pgp.com --send-key keybase.io/jlb1234

**Comment by brainwane** [comment]
[The new beta version of Enigmail, 1.9 beta 2, includes Keybase key lookup](https://sourceforge.net/p/enigmail/forum/announce/thread/6393519e/) and [is available at addons.mozilla.org](https://addons.mozilla.org/en-US/seamonkey/addon/enigmail/versions/?page=1#version-1.9b2). Thanks, Patrick Brunschwig! CC: @wsargent, @EliW, @iamale, @schock who mentioned being Enigmail users.

**Comment by johannesl** [comment]
+1

**Comment by chigh** [comment]
Mitar> But you do have them on Keybase? Can you elaborate why do you feel
like that? It could help provide some insight.

I don't have them on a key server because as it stands, they can never be
removed. Ever.  There are keys out there for me (depending upon which key
server you query because they don't all report the same results!) that are
17 years old. Some are revoked some aren't. Many that aren't revoked, the
keys are long since gone (as are the email addresses). So, there's no way
to clean that up.

Zach> This is entirely by design -- so to call it a problem is pretty,
well, problematic.

So be it. It doesn't mean I have to like it, therefore my current keys are
not on the current, inconsistent infrastructure of key servers and they
will never be submitted by me. I may agree with some of the reasoning for
it, but... preference rules here.

My other issues with them are similar to Jason's.

Zach> Personally I believe it's important for Keybase to integrate -- at
least to some extent -- with the existing public key infrastructure.

I can, however, agree with that. I can't say at this time, I where the line
is drawn.

**Comment by decibyte** [comment]
+1 from me as well.

**Comment by ryanturner** [comment]
+1

**Comment by hwmrocker** [comment]
+1

**Comment by lindhe** [comment]
> The new beta version of Enigmail, 1.9 beta 2, includes Keybase key lookup and is available at addons.mozilla.org. Thanks, Patrick Brunschwig @pbrunschwig ! CC: @wsargent, @EliW, @iamale, @schock who mentioned being Enigmail users.

So, ehm... How do I use it? It's not obvious to me. Should I choose "Import keys from URL" and provide https://keybase.io/ as the URL, or what? I'm new to this.

**Comment by terokorp** [comment]
> The new beta version of Enigmail, 1.9 beta 2, includes Keybase key lookup and is available at addons.mozilla.org
Great job but sadly that is implemented on wrong end

**Comment by lindhe** [comment]
So... That means I can't use it? I have to remove my :tada: from that comment immediately!

**Comment by brainwane** [comment]
[Enigmail 1.9.0](https://addons.mozilla.org/en-US/seamonkey/addon/enigmail/versions/?page=1#version-1.9), which [came out in February](https://www.enigmail.net/index.php/en/download/changelog#enig1.9), [included]((https://sourceforge.net/p/enigmail/bugs/search/?q=status%3Afixed+%26%26+_fixed%3A1.9.0&page=2)) the new Keybase key lookup. Note that 1.9.0 may require a GnuPG upgrade for you: "This version requires GnuPG 2.x.; GnuPG 1.4 is not supported anymore." The current version of Enigmail is 1.9.2, released 25 April 2016. 

@lindhe I also don't know quite how to use it, and I suspect there may be a bug; I've [filed an issue](https://sourceforge.net/p/enigmail/bugs/608/). I think the way it's meant to work is, if you go into your Enigmail preferences, you can add keybase.io as a keyserver. But that does not work right now, at least for me.

**Comment by DreamFlasher** [comment]
@brainwane How do I use keybase in Enigmal? Couldn't find any option in Enigmail 1.9.2 and also no info when googling. 

It's rather strange that there are ~100 people requesting keybase to provide a keyserver API for two years now, and the devs of keybase, well, idk, don't care? Have never used PGP?

**Comment by decibyte** [comment]
@DreamFlasher (and others),

Try following the link to @brainwane's Enigmail issue. There's a comment:

> You cannot use keybase.io like any other keyserver, as it does not support the hkp protocol. You have to use the keybase server from the dropdwon [sic] list (keybase://keybase.io)

I don't have a dropdown in my Enigmail UI, but a comma separated list of keyservers in a text field. I've added this there now, but can't come up with a recipient for whom I don't already have the key, to send a test e-mail and see whether the key is fetched from Keybase. So I cannot (yet) confirm whether it actually works.

**Comment by DreamFlasher** [comment]
@decibyte Thanks! I also don't have a dropdown. So I tried keybase://keybase.io as keyserver, but it doesn't work. I have one contact to test with, I wasn't able to find his public key (which he has on his public keybase.io profile) with the search for his email nor his username.

**Comment by lindhe** [comment]
Same for me, can't find any key.

![screenshot from 2016-06-10 14-01-51](https://cloud.githubusercontent.com/assets/7773090/15963663/072cb1d6-2f14-11e6-9ad3-3357bc5bf71c.png)

**Comment by DreamFlasher** [comment]
I posted in the ticket: https://sourceforge.net/p/enigmail/bugs/608/ -- apparently it's broken in the current version and supposed to be fixed in 2.0.

**Comment by lindhe** [comment]
Well, this is what we will get as long as Keybase does not get its shit together and implements some way to properly detect keybase keys as regular keys on a server (either by making it self a key server or, probably much much easier, by automatically publishing to an already established keyserver).

And this won't stop with Enigmail. EVERY SINGLE PROGRAM using online keys will have to implement their solution on top of the existing framework. This is not sustainable guys.

**Comment by DreamFlasher** [comment]
I fully agree with you @lindhe -- This also diminishes the trust I have in keybase, not integrating with keyservers (neither publishing nor acting as a keyserver) shows to me that the devs have little to no understanding of the PGP infrastructure. Furthermore this has been requested for over 2 years now by hundreds of people... Looks like they give a shit about their userbase?

**Comment by zQueal** [comment]
> This also diminishes the trust I have in keybase, not integrating with keyservers (neither publishing nor acting as a keyserver) shows to me that the devs have little to no understanding of the PGP infrastructure.

That's an incredibly pointed, unnecessary, and aggressive statement--don't you think?

> Furthermore this has been requested for over 2 years now by hundreds of people... Looks like they give a shit about their user base?

See my previous statement. Additionally, the developers _have already commented on this thread_ and fully [disclosed their plans](https://github.com/keybase/keybase-issues/issues/327#issuecomment-75983909). Having Keybase do the exact same thing as literally every other PGP key service isn't what they have in mind--and they __are__ willing to implement this. But it's not so simple and important that they're breaking their necks to get it done today. It may be on the roadmap, but it's not a priority.

Even farther, I really don't find it reasonable to assume that __every__ user would want their PGP key pushed to a Keyserver. I know I don't and I've used applications in the past which have done it automatically without my permission. That's not cool--because now to get them off the key server the key has to be revoked and now I have an unnecessary amount of work to do because it was _assumed_ that I wanted it done. That's not cool. Instead of assuming, why can't you simply do what's __always__ been done, and just execute __one__ command (once your key is already in your GPG keyring) to push to a keyserver. Incase you don't know it;

    gpg --keyserver pgp.mit.edu --send-key ********

Even if that's too difficult for you to execute you can always setup a simple alias and have this done regardless of what keybase is doing;

    function keypush(){ gpg --keyserver pgp.mit.edu --send-key ********; }

So I really don't see the need to get all uppity with the passive aggressiveness over such a "problem" which is so easily solved until it's ready to be implemented as an automatic process in the Keybase binary.....

Let's not forget that Keybase is still in __BETA__. It's not ready for normal everyday usage just yet so whining like kids about features which aren't implemented yet when there is much more to do on the roadmap is a little......childish.

**Comment by DreamFlasher** [comment]
> That's an incredibly pointed, unnecessary, and aggressive statement--don't you think?

No, I don't think so. Not complying with existing standards is either due to ignorance or stupidity. That's not an aggressive statement, that's a fact :)

Furthermore, reacting with passive-aggressiveness (explaining how pgp works, seriously? You were the one who doesn't know how it works apparently, apart from all others here ^^) or telling your users to be childish... well that's just unprofessional. 

Apart from that you really appear not to having read or understand what this is all about: This is not about people having difficulties publishing their keys on a keyserver. It is about to get public keys from keybase. Where is the difference? There are users who don't have their keys on a keyserver and only on keybase. Deliberately or not. So yes, I agree with you it shouldn't be pushed to keyservers. All what _I_ am asking for is that we can continue using our software which uses existing protocols, which is hkp, to retrieve public keys. 

Hiding behind a beta label is not convincing. Apparently more than 100 people have been asking for this within the last two years. With that argument I'd recommend you to stay in beta forever so that you can do with your users whatever you want.

**Comment by zQueal** [comment]
I don't work for Keybase. I'm not affiliated with the project at all. I don't speak for them either; so this:

> With that argument I'd recommend you to stay in beta forever so that you can do with your users whatever you want.

Is entirely uncalledfor. The only reason why my posts say "Keybase Member" is because I'm included in a group which was allowed to beta test the Go client before anyone else (to ensure stability). If I recall I was one of the only users not affiliated with the project which used Windows exclusively.

> Furthermore, reacting with passive-aggressiveness (explaining how pgp works, seriously?

I've never met you before in my life. How would I know what type of environment you're working with, or what your level of experience is? For example, I'm running Windows with bash installed, so running that command as an alias would be very simple to implement, but what if you're running Windows without bash? Or maybe you use Keybase on a work computer and can't directly interface with the GPG binary? Wouldn't it be difficult to push a square peg through a round hole? It wasn't meant to be passive aggressive--but if you take it that way, then I guess there's no helping that.

Everyone has their own views of what is or isn't bullshit, and what is or isn't productive. The only thing I've seen so far is you post your opinion and pass it off as fact. (which isn't cool, man [everything here is my opinion by the way--I'm not trying to pass it off as fact]) Keybase is not GPG. You talk about standards, but GPG pushing keys to a keyserver _is_ a standard and it already does it well. So why would Keybase take the time (which costs money) to make software which already exists? You're asking Keybase to take over for an already existing standard and then hollering when it's not a priority for the developers. Which is _super_ weird considering that Keybase is open source software. If you want it developed so badly, go do it then submit a PR or keep it private for your own usage. You're free to do that instead of calling other people names on the internet...

It's not a priority. It probably never will be a priority. Because GPG __already__ does what you want just not the Keybase binary. But so what, just make an alias in the interim...why spend dev time making 100 people happy when they could make 5000 happy with KBFS? The short sighted level of thinking here is.....staggering.

The only thing you've successfully done here is derail a pretty productive thread with direct accusations. Which is something open source development could entirely do without; having an opinion is fine, but _expecting_ others to value or respect that opinion simply because you have it? That's quite another thing entirely.

> This is not about people having difficulties publishing their keys on a keyserver. It is about to get public keys from keybase.

As far as I can tell--it's not. Because if it were you're asking for something which already exists __ubiquitously__. There are currently three ways to do it. Not only can you track a user publicly or privately then pull their key directly from Keybase (`keybase pgp pull`) but you can also __directly__ access anyone's public key using this schema:

    https://keybase.io/{username}/key.asc

so you don't even need to be using the Keybase software to access public keys and can directly import them using gpg;

    curl https://keybase.io/{username}/key.asc | gpg --import --

Not only those two, but you can view the users page on the web interface and just click the fingerprint to view their key in plain text.

> Hiding behind a beta label is not convincing.

I really don't want to respond to this because you already seem like the type of person who doesn't listen to others regardless (sorry)--but the software __is__ in beta. No one's hiding behind it. It's __incomplete software__. So you should go forward expecting it to have incomplete features or not work correctly at certain times. This is _expected_ behavior of actively developed software and isn't limited to Keybase or even a tiny handful of software.

Additionally, the only thing I've done here is reply to some semblance of an internet troll making me feel like Github comments is now Reddit (which is not a good feeling, by the way and I actually resent you a little for making it feel this way)--and I'm sure no one wants to see two idiots bicker about stupid stuff. (seeing how over 60 people are getting emails for this one stupid conversation--and if you're reading this, I really am sorry) So in the spirit of open development could we __please__ keep the comments civil and constructive? Spilling accusations against bootstrapped developers because they haven't yet implemented a feature you and a bunch of other people want as jerks or unthoughtful towards their users when the _software is still in active development_ is insanely childish and simply not called for.

Lastly, [this is an entirely sane response to this issue](https://github.com/keybase/keybase-issues/issues/327#issuecomment-142680270) and hopefully the direction the Keybase devs go. There's absolutely no reason to reinvent the wheel--especially so when it would only be as good or worse than the current wheel.

**Comment by terokorp** [comment]
> There's absolutely no reason to reinvent the wheel--especially so when it would only be as good or worse than the current wheel.

Yeah, good reason to ignore backward compatibility and require all other softwares to implement keybase way. This reminds me about one big corporation which doesn't (or didn't?) want to follow existing standards.

Beta or not, if car works only on manufacturers own roads, it may look good, but it's not really useful.

And also, sorry about email notification. I will unsubscribe this so no need to answer.

**Comment by maxtaco** [comment]
To the best of my knowledge, there is no published standard for the PGP key server protocol. RFC 4880 is totally mute on this topic. If my understanding is out of date, please let me know!

**Comment by davexunit** [comment]
+1 for this feature. People keep sending me keybase accounts and it's a real pain to import their keys because they don't also publish to *real* keyservers.

**Comment by ageis** [comment]
+1

**Comment by suzannealdrich** [comment]
I've been using Keybase and GPG Suite successfully together for some time. Since keybase lets me sync my key to my local machine, I can add user IDs to my PGP identity, upload my public key to a keyserver, generate a revocation certificate, and sign a transition statement. There's really nothing I'm doing differently with my Keybase key than any other PGP key, except I finally have some kind of identity verification via my social accounts and other Keybase users who are now tracking me. I guess I wasn't hanging out with the right crowd before Keybase, because nobody has ever validated my PGP identity until this service existed. Keyservers don't provide organic discoverability for keys published there, and there hasn't been a standard URL for retrieving people's keys off their websites. I like being able to link to my public key on Keybase from my social profiles, and feel like Keybase is increasing the scope of the PGP ecosystem at a far greater rate than any meatspace signing parties or keyserver technology has before.

**Comment by lindhe** [comment]
I agree, Keybase is great at many levels. But what many of us here have been pushing for is to integrate Keybase with the existing standards and solutions, so that both camps can feel happy about the additon of Keybase as a service. If keybase tries to pull off it's own thing it's gonna have an even harder time than if the current PGP-people are part of it. Let's face it — it's a very niche market. So let's at least get the existing crowd on board, and not start our own thing.

**Comment by compumike08** [comment]
While I do like the ability to look up users by e-mail address in Keybase in theory, in practice I have a concern that I'm not sure how to mitigate. Specifically, I'm concerned about exposing an API where anyone can search by e-mail address because it could be used by spammers and/or phishers to gather e-mail addresses. The identity proofs that Keybase currently supports for social media (e.g. Facebook, Twitter, GitHub, etc.) are all services that let you find people by username, and all of them have ways to hide your e-mail address if you want to.

Even with GitHub and GPG key signing, GitHub as the ability to let you hide your e-mail address and replace it publicly with a placeholder e-mail address of the format:
> *your_GitHub_username*@users.noreply.github.com
This placeholder e-mail address cannot actually send or receive any e-mails, however you can use that placeholder e-mail in your local Git commit author e-mail config so it shows in commits, and you can also create GPG keys with that placeholder e-mail as the key's e-mail address, and if you sign Git commits with that key and push the commits to GitHub, GitHub will recognize that placeholder e-mail address and mark the commit as being signed by a verified key (assuming of course that you already uploaded that GPG key to your GitHub account).

Facebook, Twitter, and most other social media services also offer the option to hide your e-mail address from the public. This helps protect users from having their e-mail addresses scraped by spammers, and then having their e-mail account bombarded by spam. If Keybase were to implement the ability to search by e-mail address, it could be used by spammers to collect valid e-mails. Even if Keybase only let you search by entering an e-mail and didn't display your e-mail on your profile page, spammers could still easily write scripts to very quickly try random e-mail addresses in the search and note any random e-mail address for which the search returned a match. Even worse, if that Keybase user also had other identities linked to their Keybase (e.g. Facebook, Twitter, etc.), then the spambot would then be able to crawl the linked identities and compile a very detailed profile of the person linked to the e-mail address. That would make it easy for spammers and phishers to automatically customize spam and/or phishing e-mails based on personal information from those identities to increase the odds of spam messaging getting through spam filters, as well as to make phishing e-mails more convincing.

One alternative to letting people look up users by e-mail address on Keybase might be to let users look up Keybase users by some sort of e-mail-like placeholder the way GitHub does. For example, I have a PGP key on Keybase that I use to GPG sign commits on GitHub. The key is associated with my GitHub placeholder e-mail address (compumike08@users.noreply.github.com). That way anyone who clones a public repository that has a commit signed by that key cannot get my actual e-mail address to send me spam/phishing e-mails. GitHub automatically marks commits signed by that key as "Verified" because I uploaded that PGP public key to my GitHub account and it recognizes the "compumike08@users.noreply.github.com" placeholder e-mail as being associated to my GitHub account. In addition, I have my GitHub account linked to my Keybase identity. However, if someone clones my repo locally and views a signed commit, they have no way of knowing that I'm a Keybase user and they can use Keybase to verify the signed commit. If Keybase assigned users a placeholder ID that was formatted as an e-mail address but didn't actually send or receive e-mails, users could add that Keybase placeholder address to their PGP keys. That way, if someone who isn't a Keybase user gets something signed with a user's public PGP key, then that person can see in the signature that the "e-mail address" has a keybase domain. Even if they've never heard of keybase, a quick Google search for the domain will get them to the Keybase website and read about Keybase. At that point, they could use the Keybase website's search feature to search by the Keybase placeholder e-mail address from the signed commit/file/e-mail/etc. to find the user's Keybase identity and access their verified public PGP key.

Even better, Keybase could let users have multiple placeholder e-mails as IDs, one for each of their PGP keys. That way anyone trying to access one of a user's Keybase PGP public keys could search by the relevant placeholder e-mail and go straight to the appropriate PGP public key in Keybase.

@malgorithms

**Comment by ChipWolf** [comment]
@compumike08 unlikely the API would be leveraged by spammers. As any API, it would have its rate limits and such. Besides, the way this issue is suggesting is to have user's keys searchable *by* email address, which is an entirely different mechanism than the reverse of such. Either way, you don't see standard PGP servers being bombarded with spam.
My email address is littered around the internet in plain text on various high traffuc sitea and I rarely am subject to spam, unless I register a domain without WHOIS privacy, which is rare.

**Comment by lindhe** [comment]
@compumike08 Do you think having it be opt-in would help? Making it so that every user would have to set "publish my email" to true, before anything could risk being exploited by APIs...

**Comment by compumike08** [comment]
@lindhe @ChipWolf:

@lindhe's suggestion for offering an opt-out option might be workable. You could also maybe offer an option to only allow people who you follow to look you up by e-mail address. Even if @ChipWolf is right and the risk of abuse by spammers is minimal, even the perception that it could be abused could hurt widespread adoption of Keybase (Keybase's stated goal is to eventually make effective encryption available for everyone, even those who are not tech savvy). An opt-out option might make people more comfortable with the idea.

**Comment by lindhe** [comment]
@compumike08 I just want to clarify that what I'm suggesting is an opt-in.

**Comment by compumike08** [comment]
@lindhe: Yes, I think opt-in would be fine.

**Comment by ChipWolf** [comment]
Looks like Google have jumped the gun https://security.googleblog.com/2017/01/security-through-transparency.html

No, I entirely disagree with opt-in @lindhe @compumike08. It defeats the purpose of being able to instantly verify signed content, the opt-in rate would be not nearly high enough for this to be effective.

**Comment by lindhe** [comment]
@ChipWolf I don't think an opt-out solution is any better than having it enabled for everyone. What would that accomplish? I think that if the API were to be abused, most people would notice that they get spam and get mad at Keybase for spreading their e-mail without notice, while an opt-in solution would entirely get rid of the feeling that Keybase did anything wrong.

Personally I would prefer a solution that had all or most people being searchable via e-mail (like with opt-out or always enabled), but I see why it is important to minimize the risk of people feeling violated by Keybase in the event that the API were to be abused.

**Comment by juanminm** [comment]
If we are talking about Keybase search feature, then yeah, retrieving the email is something bad. But even if the API would not allow that, you can still search for a user, get their public OpenPGP key (if any added) and retreive the email (if any addded) from the key.

**Comment by nicolasegosum** [comment]
+1

**Comment by consummo** [comment]
+1

**Comment by b3yond** [comment]
+1

**Comment by anapsix** [comment]
Made myself a lame proxy
GitHub: https://github.com/anapsix/keybase-keyserver-proxy
Docker Hub: https://hub.docker.com/r/anapsix/keybase-keyserver-proxy/

```
docker run -d --name keybase-keyserver-proxy --rm -p 11371:11371 anapsix/keybase-keyserver-proxy
sleep 3
gpg --keyserver 127.0.0.1 --recv-keys B416D3911B1D1AA0D47D5F4E5E93F7F309CAC1B2
docker kill keybase-keyserver-proxy
```

**Comment by wiktor-k** [comment]
I think it's clear at this point that this feature is not going to be implemented.

To be honest I'm not surprised:

1. If someone wants to query Keybase they have a nice [lookup API](https://keybase.io/docs/api/1.0/call/user/lookup) that even has CORS enabled,
2. If someone wants to have their key easily and securely discoverable in a "federated way" there is [Web Key Directory](https://wiki.gnupg.org/WKD) (for example [`gpg --locate-key torvalds@kernel.org`](https://www.kernel.org/category/signatures.html#using-the-web-key-directory) downloads the key from https://kernel.org/.well-known/openpgpkey/hu/pf113mfnx1f3eb1yiwhsipa91xfc7o4x),
3. "Linking" social accounts can be achieved by [just posting the fingerprint](https://jacob.hoffman-andrews.com/README/the-safe-way-to-put-a-pgp-key-in-your-twitter-bio/) or adding it to the [account bio](https://twitter.com/IOHK_Charles),
4. PGP was used to bootstrap the Keybase service but as time went on [their key model vastly diverged from PGP](https://keybase.io/blog/keybase-new-key-model).

**Comment by boydhako** [comment]
My two cents... is that keybase maps the usernames to emails that users use to register and then use that query a keyserver. 

That way one could send keys to keybase via a keyserver url (i.e. hkps.keybase.io). And if people want to use the keybase apps they create an account using the email made with the key.

Granted they'll have to address accounts that are sending multiple keys. Which I think they already got resolved.

If they stood up a keyserver with the mapping one would be more inclined to use keybase since pretty much none of the old school keyservers have API support.

**Comment by wiktor-k** [comment]
@boydhako, the problem with Keybase serving HKP endpoint is not a technical one. As various people showed here the actual endpoint code [is trivial](https://github.com/anapsix/keybase-keyserver-proxy/blob/master/server.rb).

As far as my casual skimming of the OpenPGP (standard) mailing list goes it seems the keyservers in the wild will slowly transition to the [e-mail verifying model](https://gitlab.com/sequoia-pgp/hagrid#hagrid) and that, I think, will happen faster than Keybase providing HKP endpoint.

**Comment by Cambridgeport90** [comment]
This model is new to me ... we will have to see what happens; not sure whether or not it already happened?

**Comment by wiktor-k** [comment]
@Cambridgeport90, I'm not sure if you replied to my message (the "e-mail verifying model") but if so it is already operational at https://keys.openpgp.org The server verifies e-mails (opt-in) and only lookup by e-mail or fingerprint works. It's also very fast, written in Rust. But it's *not* part of synchronizing pool and it strips third-party signatures. This is also the default keyserver for many clients such as Enigmail, GPG Tools as it's not vulnerable to key signature flooding attack.

(This is getting kind-of off-topic here so if you'd like to hear more details you can contact me via e-mail).

**Comment by adamretter** [comment]
+1

**Comment by ilf** [comment]
I would love to use keybase directly from my existing OpenPGP client, with `keyserver hkps://keybase.io` instead of having to use a browser or some additional non-generic software.

**Comment by casiar** [comment]
This is not going to happen.

**Comment by wiktor-k** [comment]
> This is not going to happen.

Agreed. For people that want OpenPGP and also want social proofs https://keyoxide.org/ is a better choice.

From my PoV Keybase is on life-support now and only bugfixes are in.

**Comment by casiar** [comment]
> 

Just a matter of time.

**Comment by aairey** [comment]
> For people that want OpenPGP and also want social proofs https://keyoxide.org/ is a better choice.


It is under AGPL 3.0 License, vendors will never adopt it like they have adopted Keybase (you can reference [Keybase profiles in Terraform](https://registry.terraform.io/modules/terraform-aws-modules/iam/aws/latest/submodules/iam-user#notes-for-keybase-users) to encrypt AWS IAM Users passwords and secrets with).

**Comment by wiktor-k** [comment]
> It is under AGPL 3.0 License

The web portal is but the standard it's based on is APL: https://ariadne.id/

Additionally the standard is distributed and doesn't really have any required server components (it's possible to do verifications on your own client machine).

Keyoxide is just a nice front end for displaying info that's in keys anyway. Alternative frontends are possible.

Just my 2 cents.

**Comment by lmlsna** [comment]
In the meantime:
 `curl -sSL "https://keybase.io/_/api/1.0/user/lookup.json?fields=public_keys&key_fingerprint=$finger" | jq ".them[].public_keys.primary.bundle" -rM` 
is equivalent to 
`gpg --keyserver "not-gonna-happen.keybase.io" --recv-key $finger`

**Comment by st-fresh** [comment]
keybase acquired by zoom
was always likely the secret intention of to amass users to be acquired and sell-out to highest bidder - never was then an interest to publish itself as a valid keyserver using the standard protocols

**Comment by Cambridgeport90** [comment]
I’ve stopped using Keybase and simply published my keys via Thunderbird, instead.

**Comment by anapsix** [comment]
> keybase acquired by zoom was always likely the secret intention of to amass users to be acquired and sell-out to highest bidder - never was then an interest to publish itself as a valid keyserver using the standard protocols

Too bad that Keybase never deployed a public keyserver, oh well.. Imho, ultimately, congrats on the acquisition - good for you!
